drivers/led_strip/tlc5971.c

```diff
@@ -0,0 +1,357 @@
+{
+	const struct tlc5971_config *cfg = dev->config;
+	struct tlc5971_data *data = dev->data;
+	uint8_t *data_buffer = data->data_buffer;
+	int count = 0;
+
+	/*
+	 * tlc5971 device order is reversed as the rgb data for the last device in the daisy chain
+	 * should be transmitted first.
+	 */
+	for (int device = (num_pixels / TLC5971_PIXELS_PER_DEVICE) - 1; device >= 0; device--) {
+		data_buffer[count++] = tlc5971_data_byte27(data->control_data);
+		data_buffer[count++] = tlc5971_data_byte26(data->control_data, data->gbc_blue);
+		data_buffer[count++] = tlc5971_data_byte25(data->gbc_blue, data->gbc_green);
+		data_buffer[count++] = tlc5971_data_byte24(data->gbc_green, data->gbc_red);
+
+		for (int pixel = (TLC5971_PIXELS_PER_DEVICE - 1); pixel >= 0; pixel--) {
+			/* data is "reversed" so RGB0 comes last, i.e at byte 0 */
+			const struct led_rgb *pixel_data =
+				&pixels[(device * TLC5971_PIXELS_PER_DEVICE) + pixel];
+
+			/*
+			 * Convert pixel data into SPI frames, mapping user colors to tlc5971
+			 * data frame color order (BGR).
+			 */
+			for (int color = 0; color < cfg->num_colors; color++) {
+				uint8_t temp =
+					tlc5971_map_color(cfg->color_mapping[color], pixel_data);
+
+				/*
+				 * The tlc5971 rgb values are 16 bit but zephyr's rgb values are
+				 * 8 bit. Simply upscale to 16 bit by using the 8 bit value for both
+				 * LSB and MSB of the 16 bit word.
+				 */
+				data_buffer[count++] = temp;
+				data_buffer[count++] = temp;
+			}
+		}
+	}
+}
+
+static int tlc5971_transmit_data(const struct device *dev, size_t num_pixels)
+{
+	const struct tlc5971_config *cfg = dev->config;
+	struct tlc5971_data *data = dev->data;
+
+	struct spi_buf buf = {
+		.buf = data->data_buffer,
+		.len = (num_pixels / TLC5971_PIXELS_PER_DEVICE) * TLC5971_PACKET_LEN,
+	};
+
+	const struct spi_buf_set tx = {
+		.buffers = &buf,
+		.count = 1,
+	};
+
+	return spi_write_dt(&cfg->bus, &tx);
+}
+
+static int tlc5971_update_rgb(const struct device *dev, struct led_rgb *pixels, size_t num_pixels)
+{
+	const struct tlc5971_config *cfg = dev->config;
+
+	if (num_pixels > cfg->num_pixels) {
+		LOG_ERR("invalid number of pixels, %zu vs actual %i", num_pixels, cfg->num_pixels);
+		return -EINVAL;
+	}
+
+	tlc5971_fill_data_buffer(dev, pixels, num_pixels);
+
+	return tlc5971_transmit_data(dev, num_pixels);
+}
+
+static int tlc5971_update_channels(const struct device *dev, uint8_t *channels, size_t num_channels)
+{
+	ARG_UNUSED(dev);
+	ARG_UNUSED(channels);
+	ARG_UNUSED(num_channels);
+
+	return -ENOTSUP;
+}
+
+int tlc5971_set_global_brightness(const struct device *dev, struct led_rgb pixel)
+{
+	const struct tlc5971_config *cfg = dev->config;
+	struct tlc5971_data *data = dev->data;
+	int res = -EINVAL;
+
+	if ((pixel.r <= TLC5971_GLOBAL_BRIGHTNESS_CONTROL_MAX) &&
+	    (pixel.g <= TLC5971_GLOBAL_BRIGHTNESS_CONTROL_MAX) &&
+	    (pixel.b <= TLC5971_GLOBAL_BRIGHTNESS_CONTROL_MAX)) {
+		data->gbc_blue = tlc5971_map_color(cfg->color_mapping[0], &pixel);
```

Comment:
The code looks correct but it is weird to have this variables named gbc_{blue,green,red). IIUC, if for example the color mapping is GBR, then the gbc_blue variable will hold the global brightness value of the green color.

Reply:
Renamed global brightness control variables to color_1, 2 and 3 with a comment about the mapping.

Reply:
Thanks. I think this makes the code more understandable. Note that this change introduces some checkpatch warnings.